File: ws/first_ws/src/service_pkg/src/client_custom_srv3.cpp
//
// Created by zcb on 2021/4/9.
//

#include <ros/ros.h>
#include <iostream>
#include <demo_srvs/FindTeam.h>

using namespace std;

int main(int argc, char **argv) {

    string nodeName = "cpp_client_node";
    ros::init(argc, argv, nodeName);
    ros::NodeHandle node;

    string serviceName = "/cpp/service";
    ros::ServiceClient client = node.serviceClient<demo_srvs::FindTeam>(serviceName);

    client.waitForExistence(); // 如果client先于 server 起来，可以一直等server

    demo_srvs::FindTeam service;
    service.request.stu.name = "jane";
    service.request.stu.age = 18;
    bool flag = client.call(service);
    if (flag) {
        ROS_INFO_STREAM("result is: " << service.response.team);
    } else {
        ROS_DEBUG_STREAM("call fail");
    }


    return 0;
}

File: 06ros/ws01/first_ws/src/service_pkg/scripts/client_custom_srv3.py
#! /usr/bin/env python
# coding:utf8

import rospy
from demo_srvs.srv import FindTeam, FindTeamRequest, FindTeamResponse

if __name__ == '__main__':
    nodeName = "python_client_node"
    rospy.init_node(nodeName)

    serviceName = "/python/service"

    # 与c++ api 不一致，
    client = rospy.ServiceProxy(serviceName, FindTeam)
    client.wait_for_service()

    # 与c++ api 不一致,这里只传 request
    request = FindTeamRequest()
    request.stu.name = "fuck"
    request.stu.age = 18

    # 与c++ api 又不一致，艹艹艹艹艹艹
    try:
        response = client.call(request)
        if isinstance(response, FindTeamResponse):
            rospy.loginfo("team's name is %s" % response.team.name)
    except Exception as e:
        print(e.message)
        print("call fail")
